python/config/series_description_matching.py

```diff
@@ -0,0 +1,180 @@
+        alma_items, logger
+    )
+    aspace_match_data, tcs_with_duplicate_keys = _get_aspace_match_data(
+        aspace_containers, logger
+    )
+
+    # find matches by comparing keys in _match_data dictionaries
+    matched_aspace_containers = []
+    for alma_key, alma_item in alma_match_data.items():
+        if alma_key in aspace_match_data:
+            tc = aspace_match_data[alma_key]
+            # get barcode from Alma item and add it to ASpace top container
+            barcode = alma_item.get("barcode")
+            tc["barcode"] = barcode
+            matched_aspace_containers.append(tc)
+
+            if logger:
+                logger.info(
+                    f"Matched item {alma_item.get('pid')} "
+                    f"with top container {tc.get('uri')}"
+                )
+
+    # find unmatched Alma items and ASpace top containers
+    alma_keys = set(alma_match_data.keys())
+    aspace_keys = set(aspace_match_data.keys())
+
+    unmatched_alma_keys = alma_keys - aspace_keys
+    unmatched_aspace_keys = aspace_keys - alma_keys
+
+    unmatched_alma_items = [alma_match_data[key] for key in unmatched_alma_keys]
+    unmatched_aspace_containers = [
+        aspace_match_data[key] for key in unmatched_aspace_keys
+    ]
+
+    # assemble unhandled data dict
+    unhandled_data = {
+        "unmatched_alma_items": unmatched_alma_items,
+        "unmatched_aspace_containers": unmatched_aspace_containers,
+        "items_with_duplicate_keys": items_with_duplicate_keys,
+        "tcs_with_duplicate_keys": tcs_with_duplicate_keys,
+    }
+
+    return matched_aspace_containers, unhandled_data
+
+
+def _get_aspace_match_data(
+    aspace_containers: list, logger: Optional[Any] = None
+) -> tuple[dict[tuple, list[tuple]]]:
+    """Parses ASpace top container indicators into indicator and series and extracts the type.
+    Returns a dictionary with the indicator, type, and series as keys, and a list of top
+    containers with duplicate keys."""
+    match_data = {}
+    tcs_with_duplicate_keys = []
+    for tc in aspace_containers:
+        tc_indicator_with_series = tc.get("indicator")
+        # if indicator begins with a number, split it into indicator and series
+        # numerical part is the indicator, the rest is the series
+        # e.g. "330M" -> indicator "330", series "M"
+        if tc_indicator_with_series[0].isdigit():
+            tc_indicator = ""
+            for char in tc_indicator_with_series:
+                if char.isdigit():
+                    tc_indicator += char
+                else:
+                    break
+            tc_series = tc_indicator_with_series[len(tc_indicator) :]
+        # if indicator does not begin with a number, also split it into indicator and series
+        # hyphen delimited
+        # e.g. "SR-130" -> indicator "130", series "SR"
+        else:
+            tc_indicator = tc_indicator_with_series.split("-")[1]
+            tc_series = tc_indicator_with_series.split("-")[0]
```

Comment:
This can be simplified somewhat, via regular expressions.  The core logic boils down to this, since there are only 2 cases to handle:
```
      if tc_indicator_with_series[0].isdigit():
          # DigitsLetter
          (tc_indicator, tc_series) = re.findall(r"(\d+)(\w+)", tc_indicator_with_series)[0]
      else:
          # Letters-Digits
          (tc_series, tc_indicator) = re.findall(r"(\w+)-(\d+)", tc_indicator_with_series)[0]
```
`re` comes from `import re`.  In both cases here, we look for 2 pieces of data:
1.  digit(s), via `\d+`
2. letter(s), via `\w+`
Each is wrapped in `{}`, to be "captured".  When multiple elements are captured like this, [re.findall()](https://docs.python.org/3/library/re.html#re.findall) returns a list of tuples.  We only want the first match, and we trust that this never returns `None` since the list indexing will fail. 

Since there's some complexity, I think it's worth wrapping this bit of code in its own function, and adding a check for `None`, at least.  Might also want to do something if multiple matches are found - I see at least one oddify in `'S-1464, S-1466 and S-1467'`.

I wrote a little stand-alone program / harness for this, to dump all the values to file for easier review.  It looks pretty clean to me but please double-check and change as needed.
```
import re
from add_alma_barcodes_to_archivesspace import _get_container_refs_from_db, get_aspace_containers
from asnake.client import ASnakeClient
from MySQLdb import connect
aspace_client = ASnakeClient(config_file=".archivessnake_secret_DEV.yml")
db_settings = aspace_client.config.get("database")
mysql_client = connect(
     host=db_settings.get("host"),
     database=db_settings.get("database"),
     user=db_settings.get("user"),
     password=db_settings.get("password"),
)

# 1273 is RKO resource
resource_id = 1273
container_refs = _get_container_refs_from_db(mysql_client, resource_id)
aspace_containers = get_aspace_containers(aspace_client, container_refs)

# DigitsLetter, or Letters-Digits, other than
# 'S-1464, S-1466 and S-1467',

with open(f"indicator_series_{resource_id}_dump.txt", "w") as f:
  for tc in aspace_containers:
      tc_indicator_with_series = tc.get("indicator")
      if tc_indicator_with_series[0].isdigit():
          # DigitsLetter
          (tc_indicator, tc_series) = re.findall(r"(\d+)(\w+)", tc_indicator_with_series)[0]
      else:
          # Letters-Digits
          (tc_series, tc_indicator) = re.findall(r"(\w+)-(\d+)", tc_indicator_with_series)[0]
      f.write(f"{tc_indicator_with_series}\t{tc_indicator}\t{tc_series}\n")
```